File: lightroom_tagger/core/catalog_sync.py
"""Incremental catalog sync — additions-only refresh from .lrcat into library.db."""

from __future__ import annotations

import os
import sqlite3
from dataclasses import dataclass
from typing import Callable

from lightroom_tagger.core.database.catalog import store_images_batch
from lightroom_tagger.lightroom.reader import connect_catalog, get_image_by_id, list_catalog_file_ids

CATALOG_LOCK_ACTIONABLE_MSG = (
    "Cannot read Lightroom catalog (locked or unavailable). "
    "Close Lightroom Classic or set LIGHTRoom_CATALOG_LOCKING_MODE=NORMAL and retry."
)


class CatalogSyncError(Exception):
    """Catalog could not be opened or read (typically locked while Lightroom is open)."""


@dataclass(frozen=True)
class CatalogSyncResult:
    added: int
    stale: int
    locking_mode: str
    catalog_total: int
    library_total: int
    missing_ids_count: int


def _current_locking_mode() -> str:
    return os.getenv("LIGHTRoom_CATALOG_LOCKING_MODE", "EXCLUSIVE").upper()


def _is_catalog_unavailable_error(exc: BaseException) -> bool:
    msg = str(exc).lower()
    return (
        "unable to open database file" in msg
        or "database is locked" in msg
        or "disk i/o error" in msg
    )


def list_library_catalog_ids(lib_db: sqlite3.Connection) -> set[int]:
    """Numeric catalog ids already indexed in ``images.id`` (TEXT column).

    Rows with empty or non-numeric ids are skipped so legacy data cannot break sync.
    """
    rows = lib_db.execute("SELECT id FROM images").fetchall()
    ids: set[int] = set()
    for row in rows:
        if isinstance(row, dict):
            raw = row.get('id')
        elif isinstance(row, sqlite3.Row):
            raw = row['id']
        else:
            raw = row[0]
        if raw is None:
            continue
        text = str(raw).strip()
        if not text:
            continue
        try:
            ids.add(int(text))
        except ValueError:
            continue
    return ids


def sync_catalog(
    catalog_path: str,
    lib_db: sqlite3.Connection,
    *,
    log: Callable[[str, str], None] | None = None,
    progress: Callable[[int, str | None], None] | None = None,
) -> CatalogSyncResult:
    """Diff catalog ids against library.db, fetch metadata for missing ids only, upsert.

    Never deletes library rows. Reports stale count (library minus catalog) for logging.
    """
    locking_mode = _current_locking_mode()

    def _log(level: str, message: str) -> None:
        if log is not None:
            log(level, message)

    try:
        catalog_conn = connect_catalog(catalog_path)
    except (sqlite3.Error, OSError) as exc:
        raise CatalogSyncError(CATALOG_LOCK_ACTIONABLE_MSG) from exc

    try:
        try:
            catalog_ids = set(list_catalog_file_ids(catalog_conn))
        except sqlite3.Error as exc:
            if _is_catalog_unavailable_error(exc):
                raise CatalogSyncError(CATALOG_LOCK_ACTIONABLE_MSG) from exc
            raise
    except CatalogSyncError:
        catalog_conn.close()
        raise

    library_ids = list_library_catalog_ids(lib_db)
    missing_ids = sorted(catalog_ids - library_ids)
    stale_count = len(library_ids - catalog_ids)

    _log(
        "info",
        (
            f"[catalog-sync] mode=set_difference locking_mode={locking_mode} "
            f"catalog_total={len(catalog_ids)} library_total={len(library_ids)} "
            f"missing={len(missing_ids)} stale={stale_count}"
        ),
    )

    records: list[dict] = []
    total_missing = len(missing_ids)
    for index, image_id in enumerate(missing_ids):
        record = get_image_by_id(catalog_conn, image_id)
        if record:
            records.append(record)
        if progress is not None and total_missing:
            pct = 5 + int(90 * (index + 1) / total_missing)
            progress(pct, f"Fetching catalog metadata {index + 1}/{total_missing}")

    added = 0
    if records:
        added = store_images_batch(lib_db, records)

    catalog_conn.close()

    if progress is not None:
        progress(100, "Catalog sync complete")

    _log(
        "info",
        (
            f"[catalog-sync] complete added={added} stale={stale_count} "
            f"locking_mode={locking_mode}"
        ),
    )

    return CatalogSyncResult(
        added=added,
        stale=stale_count,
        locking_mode=locking_mode,
        catalog_total=len(catalog_ids),
        library_total=len(library_ids),
        missing_ids_count=len(missing_ids),
    )

File: apps/visualizer/backend/src/lightroom/catalog-sync.ts
/**
 * Incremental catalog sync — additions-only refresh from a `.lrcat` into
 * `library.db`.
 *
 * The whole design is one set difference. Reading full metadata for 43,000 images
 * takes minutes; reading their ids takes one query, so the expensive join runs only
 * for the ids `library.db` has never seen. Rows are never deleted: an image missing
 * from the catalog is reported as `stale` and left alone, because the user may have
 * moved it to a different catalog and every score and description hangs off its key.
 */
import type { Db } from '../db/connection.js';
import { storeImagesBatch } from '../db/library/catalog.js';
import { libraryWrite } from '../db/library/write.js';
import {
  catalogReadonlyUriEnabled,
  connectCatalogReadOnly,
  getImageById,
  listCatalogFileIds,
  resolveCatalogLockingMode,
  type CatalogRecord,
} from './reader.js';

export const CATALOG_LOCKED_MSG =
  'Cannot read Lightroom catalog: another process holds the catalog open. ' +
  'Close Lightroom Classic and retry.';

/** The catalog could not be opened or read — typically Lightroom is holding it. */
export class CatalogSyncError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CatalogSyncError';
  }
}

export interface CatalogSyncResult {
  added: number;
  stale: number;
  locking_mode: string;
  catalog_total: number;
  library_total: number;
  missing_ids_count: number;
}

function isCatalogLockedError(e: unknown): boolean {
  return String(e instanceof Error ? e.message : e)
    .toLowerCase()
    .includes('database is locked');
}

/**
 * Turn a driver-level failure into something the UI can act on.
 *
 * Exported for its own test: reaching the locked branch through a real lock means
 * waiting out the connection's 30-second busy timeout, and these three strings are
 * the whole point of the function.
 */
export function catalogSyncErrorMessage(e: unknown): string {
  const raw = e instanceof Error ? e.message : String(e);
  const msg = raw.toLowerCase();
  if (msg.includes('database is locked')) return CATALOG_LOCKED_MSG;
  if (msg.includes('unable to open database file')) {
    return (
      'Cannot open Lightroom catalog file. On network storage, ensure the ' +
      'catalog is reachable. For legacy read-write opens, try ' +
      'LIGHTROOM_CATALOG_LOCKING_MODE=EXCLUSIVE. Close Lightroom Classic ' +
      'if the catalog is in use.'
    );
  }
  return `Cannot read Lightroom catalog: ${raw}`;
}

/**
 * The numeric catalog ids already indexed in `images.id`.
 *
 * That column is TEXT, so it can hold anything an older import wrote. Empty and
 * non-numeric values are skipped rather than raising — and the comparison has to be
 * numeric, because as text `'38887'` sorts after `'99999'` and a lexicographic diff
 * would re-fetch images that are already there.
 */
export function listLibraryCatalogIds(db: Db): Set<number> {
  const rows = db.prepare('SELECT id FROM images').all() as { id: unknown }[];
  const ids = new Set<number>();
  for (const row of rows) {
    if (row.id === null || row.id === undefined) continue;
    const text = String(row.id).trim();
    // `Number('')` is 0 and `Number('12abc')` is NaN, where Python's `int()` raises
    // for both; the regex keeps a legacy row from inventing an id of zero.
    if (!/^[+-]?\d+$/.test(text)) continue;
    ids.add(Number(text));
  }
  return ids;
}

export interface SyncCatalogOptions {
  log?: (level: string, message: string) => void;
  progress?: (pct: number, message: string) => void;
  /** Checked between metadata fetches; the run stops and reports what it added. */
  isCancelled?: () => boolean;
}

/** True when the run stopped early because the caller cancelled it. */
export interface SyncCatalogOutcome {
  result: CatalogSyncResult;
  cancelled: boolean;
}

/**
 * Diff catalog ids against `library.db`, fetch metadata for the missing ones, upsert.
 *
 * Throws `CatalogSyncError` when the catalog cannot be opened or listed. Once the
 * id list is in hand nothing else is fatal: a per-image fetch that returns nothing
 * just contributes no record.
 */
export function syncCatalog(
  catalogPath: string,
  libDb: Db,
  opts: SyncCatalogOptions = {},
): SyncCatalogOutcome {
  const log = (level: string, message: string): void => opts.log?.(level, message);

  let catalogConn: Db;
  try {
    catalogConn = connectCatalogReadOnly(catalogPath);
  } catch (e) {
    throw new CatalogSyncError(catalogSyncErrorMessage(e));
  }

  const lockingMode = resolveCatalogLockingMode(catalogReadonlyUriEnabled());

  try {
    let catalogIds: Set<number>;
    try {
      catalogIds = new Set(listCatalogFileIds(catalogConn));
    } catch (e) {
      throw isCatalogLockedError(e) ? new CatalogSyncError(CATALOG_LOCKED_MSG) : e;
    }

    const libraryIds = listLibraryCatalogIds(libDb);
    const missingIds = [...catalogIds].filter((id) => !libraryIds.has(id)).sort((a, b) => a - b);
    const staleCount = [...libraryIds].filter((id) => !catalogIds.has(id)).length;

    log(
      'info',
      `[catalog-sync] mode=set_difference locking_mode=${lockingMode} ` +
        `catalog_total=${catalogIds.size} library_total=${libraryIds.size} ` +
        `missing=${missingIds.length} stale=${staleCount}`,
    );

    const records: CatalogRecord[] = [];
    const totalMissing = missingIds.length;
    let cancelled = false;
    for (const [index, imageId] of missingIds.entries()) {
      // Python has no check here at all: `cancel_scope` is installed by the handler
      // but nothing on this path consults it, so cancelling a 43,000-image sync did
      // nothing until it finished. What it has already fetched is still written.
      if (opts.isCancelled?.()) {
        cancelled = true;
        break;
      }
      const record = getImageById(catalogConn, imageId);
      if (record) records.push(record);
      if (totalMissing) {
        const pct = 5 + Math.trunc((90 * (index + 1)) / totalMissing);
        opts.progress?.(pct, `Fetching catalog metadata ${index + 1}/${totalMissing}`);
      }
    }

    const added = records.length ? libraryWrite(libDb, () => storeImagesBatch(libDb, records)) : 0;

    opts.progress?.(100, 'Catalog sync complete');
    log(
      'info',
      `[catalog-sync] complete added=${added} stale=${staleCount} ` +
        `locking_mode=${lockingMode}`,
    );

    return {
      result: {
        added,
        stale: staleCount,
        locking_mode: lockingMode,
        catalog_total: catalogIds.size,
        library_total: libraryIds.size,
        missing_ids_count: totalMissing,
      },
      cancelled,
    };
  } finally {
    catalogConn.close();
  }
}
